hbase-server/src/test/java/org/apache/hadoop/hbase/quotas/TestScannerLeaseCount.java

```diff
@@ -0,0 +1,215 @@
+import java.util.concurrent.atomic.AtomicLong;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.HBaseTestingUtil;
+import org.apache.hadoop.hbase.HConstants;
+import org.apache.hadoop.hbase.SingleProcessHBaseCluster;
+import org.apache.hadoop.hbase.StartTestingClusterOption;
+import org.apache.hadoop.hbase.TableName;
+import org.apache.hadoop.hbase.client.ColumnFamilyDescriptorBuilder;
+import org.apache.hadoop.hbase.client.Connection;
+import org.apache.hadoop.hbase.client.ConnectionFactory;
+import org.apache.hadoop.hbase.client.ResultScanner;
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.hadoop.hbase.client.Table;
+import org.apache.hadoop.hbase.client.TableDescriptorBuilder;
+import org.apache.hadoop.hbase.regionserver.LeaseManager;
+import org.apache.hadoop.hbase.regionserver.Region;
+import org.apache.hadoop.hbase.regionserver.RegionServerServices;
+import org.apache.hadoop.hbase.testclassification.ClientTests;
+import org.apache.hadoop.hbase.testclassification.MediumTests;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.apache.hadoop.hbase.util.JVMClusterUtil;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+import org.apache.hadoop.hbase.shaded.protobuf.generated.ClientProtos;
+
+@Category({ MediumTests.class, ClientTests.class })
+public class TestScannerLeaseCount {
+  private static final HBaseTestingUtil UTIL = new HBaseTestingUtil();
+  private static final TableName TABLE_NAME = TableName.valueOf("ScannerLeaseCount");
+  private static final byte[] FAM = Bytes.toBytes("Fam");
+
+  private static volatile boolean SHOULD_THROW = false;
+
+  private static Connection CONN;
+  private static Table TABLE;
+
+  private static SingleProcessHBaseCluster cluster;
+
+  @BeforeClass
+  public static void setUp() throws Exception {
+    StartTestingClusterOption option =
+      StartTestingClusterOption.builder().rsClass(MockedQuotaManagerRegionServer.class).build();
+    cluster = UTIL.startMiniCluster(option);
+    UTIL.getAdmin().createTable(TableDescriptorBuilder.newBuilder(TABLE_NAME)
+      .setColumnFamily(ColumnFamilyDescriptorBuilder.of(FAM)).build());
+    Configuration conf = new Configuration(UTIL.getConfiguration());
+    CONN = ConnectionFactory.createConnection(conf);
+    TABLE = CONN.getTable(TABLE_NAME);
+    UTIL.loadTable(TABLE, FAM);
+  }
+
+  @AfterClass
+  public static void tearDown() throws Exception {
+    try {
+      TABLE.close();
+    } catch (Exception ignore) {
+    }
+    try {
+      CONN.close();
+    } catch (Exception ignore) {
+    }
+    UTIL.shutdownMiniCluster();
+  }
+
+  @Before
+  public void before() {
+    SHOULD_THROW = false;
+  }
+
+  @Test
+  public void itIncreasesScannerCount() throws Exception {
+    long beforeScan = countLeases();
+    ResultScanner scanner = TABLE.getScanner(new Scan());
+    Thread.sleep(1000);
```

Comment:
We need to wait until the scan and lease are created server-side. Otherwise, our lease counting will not reflect this new scan that was created

Reply:
Tests with fixed timed waits like this are guaranteed to show up on our flakey dashboard. Instead of this, I would loop the `beforeScan < afterScan` in a retry-with-timeout structure. We have the `Waiter` class and `HBaseCommonTestingUtil#waitFor` methods for exactly this purpose. Whatever hacks you required, make sure you include a comment in code, not just on the PR.